pulser-simulation/pulser_simulation/qutip_backend.py

```diff
@@ -88,3 +96,138 @@ def run(
+            noise_model = sequence.device.default_noise_model
+        noise_model = noise_model or self._config.noise_model
+        simconfig = SimConfig.from_noise_model(noise_model)
+        self._sim_obj = QutipEmulator.from_sequence(
+            sequence,
+            sampling_rate=self._config.sampling_rate,
+            config=simconfig,
+            with_modulation=self._config.with_modulation,
+        )
+        self._sim_obj.set_evaluation_times(
+            self._config._get_legacy_evaluation_times(
+                self._sim_obj.total_duration_ns
+            ),
+        )
+        if self._config.initial_state:
+            self._sim_obj.set_initial_state(
+                self._config.initial_state.to_qobj()
+            )
+
+    def run(self) -> Results:
+        """Executes the sequence on the backend."""
+        res = Results(
+            atom_order=tuple(self._sequence.qubit_info),
+            total_duration=self._sim_obj.total_duration_ns,
+        )
+        eigenstates = self._sim_obj.samples_obj.eigenbasis
+
+        if (
+            "amplitude" not in self._sim_obj.config.noise
+            or self._sim_obj.config.amp_sigma == 0.0
+        ) and (
+            "SPAM" not in self._sim_obj.config.noise
+            or self._sim_obj.config.eta == 0
+        ):
+            # A single run is needed, regardless of self.config.runs
+            _sim_res = self._sim_obj._run_solver(progress_bar=False)
+            assert isinstance(_sim_res, CoherentResults)
+
+            for qutip_res in _sim_res:
+                t_ = qutip_res.evaluation_time
+                state = QutipState(qutip_res.state, eigenstates=eigenstates)
+                ham: QutipOperator = QutipOperator(
+                    self._sim_obj.get_hamiltonian(t_ * res.total_duration),
+                    eigenstates=eigenstates,
+                )
+                for obs in self._config.observables:
+                    obs(
+                        config=self._config,
+                        t=t_,
+                        state=state,
+                        hamiltonian=ham,
+                        result=res,
+                    )
+
+        else:
+            density_matrices: dict[float, qutip.Qobj] = {}
+            total_reps = 0
+            for cleanres_noisyseq, reps in self._sim_obj._noisy_runs(
+                progress_bar=False
+            ):
+                total_reps += reps
+                for index, qutip_res in enumerate(cleanres_noisyseq):
+                    t_ = qutip_res.evaluation_time
+
+                    density_matrices.setdefault(
+                        t_,
+                        qutip.tensor(
+                            [
+                                qutip.Qobj(np.zeros((2, 2)))
+                                for _ in range(
+                                    self._sim_obj._hamiltonian._size
+                                )
+                            ]
```

Comment:
nit: Is this going to be executed regardless of whether `t_` already exists in `density_matrices`? If so, the old-school `if t_ in density_matrices` or a `defaultdict` would be preferable